core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableTopNHeapSort.java

```
@@ -0,0 +1,162 @@
+ * {@link org.apache.calcite.adapter.enumerable.EnumerableConvention enumerable calling convention}.
+ * It also deals with the offset and the limit.
+ */
+public class EnumerableTopNHeapSort extends Sort implements EnumerableRel {
+
+  /**
+   * Creates an EnumerableTopNHeapSort.
+   *
+   * <p>Use {@link #create} unless you know what you're doing.
+   */
+  public EnumerableTopNHeapSort(
+      RelOptCluster cluster,
+      RelTraitSet traitSet,
+      RelNode input,
+      RelCollation collation,
+      RexNode offset,
+      RexNode fetch) {
+    super(cluster, traitSet, input, collation, offset, fetch);
+    assert this.getConvention() instanceof EnumerableConvention;
+    assert this.getConvention() == input.getConvention();
+    assert fetch != null;
+  }
+
+  /** Creates an EnumerableTopNHeapSort. */
+  public static EnumerableTopNHeapSort create(
+      RelNode child,
+      RelCollation collation,
+      RexNode offset,
+      RexNode fetch) {
+    final RelOptCluster cluster = child.getCluster();
+    final RelTraitSet traitSet = cluster.traitSetOf(EnumerableConvention.INSTANCE).replace(
+        collation);
+    return new EnumerableTopNHeapSort(cluster, traitSet, child, collation, offset, fetch);
+  }
+
+  private static Expression getExpression(RexNode offset) {
+    if (offset instanceof RexDynamicParam) {
+      final RexDynamicParam param = (RexDynamicParam) offset;
+      return Expressions.convert_(
+          Expressions.call(
+              DataContext.ROOT,
+              BuiltInMethod.DATA_CONTEXT_GET.method,
+              Expressions.constant("?" + param.getIndex())), Integer.class);
+    } else {
+      return Expressions.constant(Integer.valueOf(RexLiteral.intValue(offset)));
+    }
+  }
+
+  @Override public EnumerableTopNHeapSort copy(
+      RelTraitSet traitSet,
+      RelNode newInput,
+      RelCollation newCollation,
+      RexNode offset,
+      RexNode fetch) {
+    if (fetch == null) {
+      return null;
```

I think this is breaking the contract of `copy` method. Maybe it should throw an exception instead?

---

I don't recall the contract of `copy`, but where's the harm in allowing `EnumerableTopNHeapSort` to not have a limit? Heap-sort is a valid sort algorithm, after all.